sklego/pandas_utils.py

```diff
@@ -0,0 +1,152 @@
+import inspect
+import logging
+import numpy as np
+import pandas as pd
+import datetime as dt
+
+from functools import wraps
+from sklego.common import as_list
+from scipy.ndimage.interpolation import shift
+
+
+def log_step(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        logger = logging.getLogger(inspect.stack()[1].function)
+
+        tic = dt.datetime.now()
+        result_df = func(*args, **kwargs)
+        time_taken = str(dt.datetime.now() - tic)
+        func_args = inspect.signature(func).bind(*args, **kwargs).arguments
+        func_args_str = ''.join(', {} = {!r}'.format(*item) for item in list(func_args.items())[1:])
+
+        logger.info(
+            f"[ {func.__name__}(df{func_args_str}) ] "
+            f"n_obs={result_df.shape[0]} n_col={result_df.shape[1]} time={time_taken}")
+        return result_df
+    return wrapper
+
+
+def add_lags(X, cols, lags):
+    """
+    Appends lag column(s).
+
+    :param X: array-like, shape=(n_columns, n_samples,) training data.
+    :param cols: column name(s) or index (indices).
+    :param lags: the amount of lag for each col.
+    :returns: ``pd.DataFrame | np.ndarray`` with only the selected cols.
+
+    :Example:
+
+    >>> import pandas as pd
+    >>> df = pd.DataFrame([[1, 2, 3],
+    ...                    [4, 5, 6],
+    ...                    [7, 8, 9]],
+    ...                    columns=['a', 'b', 'c'],
+    ...                    index=[1, 2, 3])
+
+    >>> add_lags(df, 'a', [1]) # doctest: +NORMALIZE_WHITESPACE
+       a  b  c  a1
+    1  1  2  3  4.0
+    2  4  5  6  7.0
+
+    >>> add_lags(df, ['a', 'b'], 2) # doctest: +NORMALIZE_WHITESPACE
```

Comment:
3. @MBrouns what is better base behaviour if we have lags=2 that we then only put in the second lag or when lags=2 we get [1 ... 2]?

Reply:
I think with `lags=2` we should only give the second lag. My reasoning for this would be that you can then use this to simulate data becoming available at a later moment.

Reply:
@koaning it's a nice functionality, but kind of unintuitive. You can produce the same result with `add_lags(df, ['a', 'b'], range(2))` in this implementation.

@MBrouns this result in the current implementation can achieved by passing `[2]`

Reply:
let keep this behavior for now at least. it's kind of an R thing to expect this behaviour but i agree this is more explicit.